View-only dav plugin using IShare attributes

mrow4a opened the pull request:
- [x] Implement extra share permissions with IShare from Approach 3 here - https://github.com/owncloud/core/issues/33458
- [x] Implement view-only plugin using extra share permissions to restrict view-only permission for read-only files.
- [x] Allow apps to register extra share permissions in frontend
- [x] Clean the code
- [x] Add unit tests
- [x] Try JSON type in DB @DeepDiver1975 
- [x] Make sure PR works with custom groups

Would be also good to merge UX for Forbidden exception with view-only https://github.com/owncloud/core/pull/33992

@PVince81 @DeepDiver1975 @pmaier1

mrow4a commented:
In PR for view-only plugin & extra share permission, I initially took an approach (https://github.com/owncloud/core/issues/33458) of extending `oc_share` table with new string column (json) `extra_share_permissions`. However, I think of going with the approach of external tables. These are pros&cons:

Extended `oc_share` table:
- currently adds new function `setExtraPermissions/getExtraPermissions` only to Share20\Share class - https://github.com/owncloud/core/pull/33994/files#diff-a19a495d2d4616d9b451f849126013a4. IShare is not touched as it would require all apps to adjust.
- easy to implement, no additional queries needed (single query retrieves the information). Expensive as each share need to reserve string space in DB
- requires to add additional API functions to IShare?, or keep with appraoch of `if instanceof Share20\Share` https://github.com/owncloud/core/pull/33994/files#diff-faddd9984dfaf09a2b85d8dc4867207aR887

New `extra_share_permissions` table:
- ExtraShareManager used to retrieve extra share permissions for shares
-`IShare` or `Share/IManager` not extended, new classes used (`Share/ExtraPermissions/IManager`)
- new table, reserves space in DB only for shares which have extra share permissions (by full-share-id/share-id).
- might need some additional queries to DB to retrieve extra share permissions. 

@DeepDiver1975 @PVince81 do you have any input?

DeepDiver1975 commented on core/Migrations/Version20181220085457.php:
give it a try with the JSON type as discussed and let's see ...

DeepDiver1975 commented on lib/private/Share20/DefaultShareProvider.php:
```diff
@@ -141,11 +142,16 @@ public function create(\OCP\Share\IShare $share) {
+		// No extra share permissions are set during share create
+		// Setting extra share permissions is only allowed at update
+		$shareExtraPermissions = new ExtraSharePermissions();
+		$qb->setValue('extra_permissions', $qb->createNamedParameter($shareExtraPermissions->serialize()));
```
the table defines a default value - no need to set anything here from my pov ....

DeepDiver1975 commented on lib/private/Share20/ExtraPermissions/Permissions.php:
```diff
@@ -0,0 +1,68 @@
+
+use OCP\Share\ExtraPermissions\IPermissions;
+
+class Permissions implements IPermissions {
```
I don't see the need of this class. Just use an array with the app as key and the permission as value

DeepDiver1975 commented on lib/private/Share20/ExtraPermissions/Permission.php:
```diff
@@ -0,0 +1,47 @@
+class Permission implements IPermission {
+
+	/** @var string */
+	private $app;
```
why app again? app is the key in the permissions array ....

mrow4a committed:
Add support for extra share attributes

Added additional share attributes for shares as an extension point for
apps to provide their own extra attributes that can be used as
extra permissions.

DeepDiver1975 commented on apps/dav/lib/DAV/ViewOnlyPlugin.php:
```diff
@@ -0,0 +1,114 @@
+/**
+ * @author Piotr Mrowczynski piotr@owncloud.com
+ *
+ * @copyright Copyright (c) 2018, ownCloud GmbH
```
2019

DeepDiver1975 commented on apps/dav/lib/DAV/ViewOnlyPlugin.php:
```diff
@@ -0,0 +1,114 @@
+	/**
+	 * ViewOnlyPlugin plugin
+	 */
+	public function __construct() {
```
remove if not needed

DeepDiver1975 commented on apps/dav/lib/DAV/ViewOnlyPlugin.php:
```diff
@@ -0,0 +1,114 @@
+			if (!$node->isUpdateable() && $canDownload !== null && !$canDownload) {
+				throw new Forbidden('File is in secure-view mode and cannot be directly downloaded.');
+			}
+		} catch (NotFound $e) {
```
log them all

DeepDiver1975 commented on apps/dav/lib/DAV/ViewOnlyPlugin.php:
```diff
@@ -0,0 +1,114 @@
+			// Check if read-only and on whether permission can download is both set and disabled.
+			$canDownload = $share->getExtraPermissions()->getPermission('dav', 'can-download');
+			if (!$node->isUpdateable() && $canDownload !== null && !$canDownload) {
+				throw new Forbidden('File is in secure-view mode and cannot be directly downloaded.');
```
translation necessary? @PVince81 do we translate exception messages in dav plugins at all?

mrow4a replied:
Not, there are no - also Dav Browser Error plugin is broken and displays legacy page, see - 
https://github.com/owncloud/core/pull/33992

DeepDiver1975 commented on apps/dav/tests/unit/DAV/ViewOnlyPluginTest.php:
```diff
@@ -0,0 +1,137 @@
+
+	public function setUp() {
+		$this->plugin = new ViewOnlyPlugin();
+		$this->request = $this->getMockBuilder('Sabre\HTTP\RequestInterface')->getMock();
```
use ::class

DeepDiver1975 replied:
use createMock

DeepDiver1975 commented on apps/dav/tests/unit/DAV/ViewOnlyPluginTest.php:
```diff
@@ -0,0 +1,137 @@
+
+	public function testCanGetNonDav() {
+		$this->request->expects($this->exactly(1))->method('getPath')->willReturn('files/test/target');
+		$this->tree->method('getNodeForPath')->willReturn('this is not Dav File');
```
getNodeForPath will not return this string - right?

DeepDiver1975 commented on apps/dav/tests/unit/DAV/ViewOnlyPluginTest.php:
```diff
@@ -0,0 +1,137 @@
+		$davNode = $this->createMock(DavFile::class);
+		$this->tree->method('getNodeForPath')->willReturn($davNode);
+
+		$davNode->method('getNode')->willReturn('this is not File or Folder');
```
get node is returning a node instance - not a string

DeepDiver1975 commented on apps/dav/tests/unit/DAV/ViewOnlyPluginTest.php:
```diff
@@ -0,0 +1,137 @@
+		$this->assertTrue($this->plugin->checkViewOnly($this->request));
+	}
+
+	public function nodeReturns() {
```
name proposal: providesDataForCanGet

DeepDiver1975 commented on apps/dav/tests/unit/DAV/ViewOnlyPluginTest.php:
```diff
@@ -0,0 +1,137 @@
+	/**
+	 * @dataProvider nodeReturns
+	 */
+	public function testCanGet($node, $canDownloadPerm, $isUpdatable, $expected) {
```
expected argument is usually always the first

DeepDiver1975 commented on apps/files_sharing/lib/Controller/Share20OcsController.php:
```diff
@@ -219,6 +221,23 @@ protected function formatShare(IShare $share, $received = false) {
 		$result['mail_send'] = $share->getMailSend() ? 1 : 0;
 
+		$extraPermissions = $share->getExtraPermissions();
+		if ($extraPermissions !== null) {
```
looks like code duplication from below - line 1201++

DeepDiver1975 commented on apps/files_sharing/tests/ApiTest.php:
```diff
@@ -1000,21 +1000,29 @@ public function testUpdateShare() {
+		$this->assertEquals(1, $share2->getPermissions());
 
 		// update permissions
 		$params = [];
```
easier to read
```php
$params = [
  'permissions' => '1',
  'extraPermissions' => [
    'app' => 'app1'
    ....
]
]

DeepDiver1975 commented on core/Migrations/Version20181220085457.php:
```diff
@@ -0,0 +1,30 @@
+<?php
+namespace OC\Migrations;
```
file header is missing

DeepDiver1975 commented on core/js/shareitemmodel.js:
```diff
@@ -752,7 +828,114 @@
+			/** @type OC.Share.Types.ShareInfo **/
+			var share = this.get('shares')[shareIndex];
+			if(!_.isObject(share)) {
+				throw "Unknown Share";
```
will the user see this? translate then

mrow4a replied:
It is everywhere in the code like that, would require refactor to change it. Anyways it is rather a signal of the bug rather than expected exception

DeepDiver1975 commented on lib/private/Share20/ExtraPermissions.php:
```diff
@@ -0,0 +1,71 @@
+	/** @var array */
+	private $permissions;
+
+	public function __construct() {
```
remove - not needed

DeepDiver1975 commented on lib/public/Share/IShare.php:
```diff
@@ -184,6 +184,30 @@ public function setPermissions($permissions);
+	 * @since 10.2.0
+	 * @return IShare The modified object
+	 */
+	public function setExtraPermissions($permissions);
```
add type to argument please

DeepDiver1975 commented on lib/public/Share/IShare.php:
```diff
@@ -184,6 +184,30 @@ public function setPermissions($permissions);
+	 * @return IExtraPermissions
+	 * @since 10.2.0
+	 */
+	public function getExtraPermissions();
```
add return type please

DeepDiver1975 reviewed (changes requested):
Looks good - this can be merged after comments have been addressed

DeepDiver1975 commented on apps/dav/lib/DAV/ViewOnlyPlugin.php:
```diff
@@ -0,0 +1,115 @@
+	/** @var \Sabre\DAV\Server $server */
+	private $server;
+
+	/** @var \Sabre\DAV\Server $server */
```
ilogger

DeepDiver1975 commented on apps/files/download.php:
```diff
@@ -39,6 +39,17 @@
+$event = new \Symfony\Component\EventDispatcher\GenericEvent(null, ['path' => $filename]);
+OC::$server->getEventDispatcher()->dispatch('file.beforeGetDirect', $event);
+if ($event->hasArgument('errorMessage')) {
+	\header("HTTP/1.0 500 Forbidden");
```
500? forbiddeb?

DeepDiver1975 commented on apps/files_sharing/lib/ViewOnly.php:
```diff
@@ -0,0 +1,110 @@
+				// access to filecache is expensive in the loop
+				$fileInfo = Filesystem::getFileInfo($file);
+				if (!$this->checkFileInfo($fileInfo)) {
+					$canDownload = false;
```
maybe return right away?

DeepDiver1975 commented on apps/files_sharing/lib/ViewOnly.php:
```diff
@@ -0,0 +1,110 @@
+		}
+		// If any of elements cannot be downloaded, prevent whole download
+		$canDownload = true;
+		$files = Filesystem::getDirectoryContent($dir);
```
hard to mock - use non-static

mrow4a replied:
@DeepDiver1975 what non-static equivalent function do you suggest?

DeepDiver1975 commented on apps/files_sharing/tests/FileListenersTest.php:
```diff
@@ -0,0 +1,212 @@
+		parent::setUp();
+		\OC_Hook::clear();
+
+		$this->eventDispatcher = \OC::$server->getEventDispatcher();
```
mock this please

mrow4a replied:
@DeepDiver1975 this was the problem I tried to solve for a long time working on unit tests for this feature. It is non trivial to test that Hook class which uses that ViewOnly class (which uses Filesystem directly).

mrow4a replied:
I can probably test ViewOnly separatelly and with Hooks class I just tests that these functions are being called - maybe this could simplify it.

DeepDiver1975 commented on apps/files_sharing/tests/FileListenersTest.php:
```diff
@@ -0,0 +1,212 @@
+		$this->eventDispatcher = \OC::$server->getEventDispatcher();
+
+		$this->hooks = new Hooks(
+			\OC::$server->getRootFolder(),
```
mock

mrow4a replied:
I had to use real classes not to make other unit tests fail (it uses global event dispatcher, and thus the registered listeners are active across many tests.. thus real classes need to be used not mocks)

DeepDiver1975 replied:
it should be possible to mock dispatcher as well as all other instances you are passing into the hooks class ... as long as no static file system methods are used (this should be avoided in any case)

jvillafanez commented on lib/private/Share20/DefaultShareProvider.php:
```diff
@@ -1238,4 +1262,47 @@ private function validate($share) {
+	 * @return IShare modified share
+	 */
+	private function updateShareAttributes(IShare $share, $data) {
+		if ($data !== null) {
```
being a private function and stating that the "$data" parameter is a string (in the phpdoc), this check doesn't seem needed. I'd move the responsability to the caller.

jvillafanez commented on lib/private/Share20/Manager.php:
```diff
@@ -1543,4 +1551,23 @@ public function sharingDisabledForUser($userId) {
+		}
+
+		$formattedPermissions = [];
+		foreach ($perms->getScopes() as $scope) {
```
I've seen this piece of code in several places... maybe it should be part of the `IAttributes`. A `toArray()` might be interesting. Note that the definition should be flexible

mrow4a replied:
@jvillafancez not sure what is the intention here

mrow4a replied:
you want IAttributes to have function toArray which returns map with attributes?

jvillafanez replied:
I think the intention here is to create a json based on the contents of the attribute object. While returning a json might be too much, just returning a known array structure that can be easily feed into a json can be enough. This piece of code can be moved into the IAttribute, which additionally will prevent code duplication.
You'll still need to do the loops, but we move the responsability to create a known array structure to the IAttribute class.

DeepDiver1975 commented on apps/files_sharing/lib/AppInfo/Application.php:
```diff
@@ -154,12 +154,19 @@ function () use ($c) {
 				$c->getServer()->getShareManager(),
-				$c->query(NotificationPublisher::class)
+				$c->query(NotificationPublisher::class),
+				$uid
```
Absolutly - furthermore it is a hard requirement in any case because of the initialization process and when the current user is know ...

mrow4a replied:
makes sense, adjusted

mrow4a commented on apps/files_sharing/lib/ViewOnly.php:
```diff
@@ -0,0 +1,116 @@
+					}
+				}
+			} catch (NotFoundException $e) {
+				continue;
```
It has unit tests, and is 100% covered for each of the cases

mrow4a replied:
Check HooksTest

jvillafanez commented on apps/files_sharing/tests/HooksTest.php:
```diff
@@ -171,4 +186,151 @@ public function testDiscardShareNotification() {
+		$senderUserFolder = $this->createMock(Folder::class);
+		$senderUserFolder->method('get')->willReturn($senderFile);
+
+		$return[] = [ '/bar.txt', $senderUserFolder, true ];
```
if possible, better use "result" as variable name to prevent any confusion

mrow4a commented on apps/dav/tests/unit/DAV/ViewOnlyPluginTest.php:
```diff
@@ -0,0 +1,132 @@
+
+		try {
+			// with these permissions / with this type of node user can download
+			$ret = $this->plugin->checkViewOnly($this->request);
```
I now assert for exception

mrow4a commented:
@jvillafanez @DeepDiver1975 adjusted to reviews

PVince81 commented on apps/files_sharing/lib/Hooks.php:
```diff
@@ -131,6 +149,64 @@ function (GenericEvent $event) {
+
+				// Check only for user/group shares. Don't restrict e.g. share links
+				if ($uid = $this->getCurrentUserUid()) {
+					$viewOnlyHandler = new ViewOnly(
```
weird that "ViewOnly" is not imported anywhere ?!

PVince81 replied:
never mind, same namespace

PVince81 commented on apps/files_sharing/lib/Hooks.php:
```diff
@@ -131,6 +149,64 @@ function (GenericEvent $event) {
+				}
+
+				// Check only for user/group shares. Don't restrict e.g. share links
+				if ($uid = $this->getCurrentUserUid()) {
```
bad practice, please use strict comparison.

in PHP `if  "0"` returns false, so user ids like "0" would fail this test...

PVince81 commented on core/Migrations/Version20181220085457.php:
```diff
@@ -0,0 +1,50 @@
+			if (!$shareTable->hasColumn('attributes')) {
+				$shareTable->addColumn(
+					'attributes',
+					Type::JSON,
```
do we have integration tests covering this ? I hope this works for all our supported DBs

DeepDiver1975 replied:
unit tests are covering this - at least as we initially discussed using this data type

PVince81 commented on core/js/sharedialogshareelistview.js:
```diff
@@ -279,7 +332,19 @@
+				attributes.push({
+					scope : $(checkbox).data('scope'),
+					key: $(checkbox).attr('name'),
+					enabled: checked
```
use the "value" attribute to make it reversible ? we can then set "value='false'" to have the checkbox negate a permission when ticked

this would open the door for more understandable checkbox labels like "restricted mode" or something instead of "can view"

PVince81 replied:
or add an attribute "reverse" in the share attribute spec

PVince81 commented on core/js/shareitemmodel.js:
```diff
@@ -176,13 +229,76 @@
+				// e.g. because of share permission/attr change
+				// and need to be initialized with their default value
+				if (!found) {
+					filteredAttributes.push({
```
adding generic attributes on the server side was ok for storage purposes, but treating them in a generic way on the client just for 2 new attributes instead of hard-coding feels a bit over-engineered.

also this will cost way more time to add unit tests to make sure your generic logic works correctly while with a few hard-coded states for 2 attributes would produce smaller tests

let's leave this as is for now...

PVince81 commented on core/js/shareitemmodel.js:
```diff
@@ -752,7 +880,98 @@
+			}
+
+			// Add attributes for this share
+			var currentAttributes = JSON.parse(share.attributes);
```
hmm, I'd have expected these to be pre-parsed when returned in the JSON response

anyway, let's leave this

PVince81 commented on lib/private/Share20/Manager.php:
```diff
@@ -641,6 +642,7 @@ public function createShare(\OCP\Share\IShare $share) {
 			'shareType' => $share->getShareType(),
 			'uidOwner' => $share->getSharedBy(),
 			'permissions' => $share->getPermissions(),
+			'attributes' => $share->getattributes(),
```
capital A

PVince81 commented on lib/public/Share/IAttributes.php:
```diff
@@ -0,0 +1,68 @@
+ * @package OCP\Share
+ * @since 10.2.0
+ */
+interface IAttributes {
```
IShareAttributes would have been better... we started a bad trend by having interface names relying on their namespaces... general confusion then when importing multiple of them with `use`

PVince81 reviewed (changes requested):
some minor changes like bad condition, typo

one nice to have: ability to switch a checkbox to set a permission to "false" instead of "true"

PVince81 commented:
also:

- [ ] missing JS tests especially for the complex frontend attribute logic

PVince81 commented:
- [ ] raise docs ticket and provide documentation for registering custom share attributes and their details

PVince81 commented on apps/dav/lib/DAV/ViewOnlyPlugin.php:
```diff
@@ -0,0 +1,115 @@
+use OCA\DAV\Connector\Sabre\Exception\Forbidden;
+use OCA\DAV\Connector\Sabre\File as DavFile;
+use OCA\DAV\Meta\MetaFile;
+use OCA\Files_Sharing\SharedStorage;
```
this will fail to import if the "files_sharing" app is disabled. better use the string instead for instance check like we do in other places.

PVince81 commented on apps/dav/lib/DAV/ViewOnlyPlugin.php:
```diff
@@ -0,0 +1,115 @@
+			}
+
+			$storage = $node->getStorage();
+			if (!$storage->instanceOfStorage(SharedStorage::class)) {
```
use string here because class might not be loaded when "files_sharing" is disabled

PVince81 commented on core/js/sharedialogshareelistview.js:
```diff
@@ -66,10 +66,18 @@
+			'<div class="shareAttributes"">' +
+			'{{#each shareAttributes}}' +
+			'<span class="shareOption">' +
+			'<input id="can-{{name}}-{{cid}}-{{shareWith}}" type="checkbox" name="{{name}}" class="attributes checkbox" {{#if enabled}}checked="checked"{{/if}} data-scope="{{scope}}" data-enabled="{{enabled}}""/>' +
```
"enabled" is semantically the same as "checked".

in frontend terms we usually use "enabled" and "disabled" to gray out a component to make it read-only for example.

so when reading this code I was confused that "enabled" actually meant "checked"

PVince81 commented:
what happens if two apps register the same attribute in the frontend ?

PVince81 commented:
fixed minor points, rebased and adjusted for php unit 6

PVince81 commented on core/js/shareitemmodel.js:
```diff
@@ -56,6 +57,25 @@
+	 * @property {string} description
+	 * @property {number[]} shareType
+	 * @property {number[]} incompatiblePermissions
+	 * @property {OC.Share.Types.ShareAttribute[]} incompatibleAttributes
```
please add the actual doc text, what are these attributes about ?

LinneyS replied:
How and where is the `description` property used?

DeepDiver1975 reviewed (changes requested):
will submit fixes for the comments asap

mrow4a committed:
Add View-only share permission using share attributes

Added share attribute "download" that can be used to allow/restrict
downloading of shared files or folders.

PVince81 commented:
I've squashed everything into two distinct commits:
1) first commit adds support for extra share permissions
2) second commit uses said support for the view-only "download" permission

PVince81 commented:
I'm a bit concerned that the JSON field might not work in all databases with upsert from this upcoming PR https://github.com/owncloud/core/pull/34769.

we will see after merging and rebasing the other one...

PVince81 commented:
stable10: https://github.com/owncloud/core/pull/34951

PVince81 commented:
seems no issues with upsert and JSON: https://github.com/owncloud/core/pull/34952

DeepDiver1975 commented:
THX @LinneyS - mind opening a pull request? We can much easier discuss this then. THX a lot!

mrow4a commented:
@LinneyS what is such permission about? We need to be sure that it follows the logic for "permissions in owncloud". That means, that checked permission box adds some higher level of permission, and unchecked lowers the permission level.

phil-davis commented:
`backport-request`  label removed. The backport was done already, see https://github.com/owncloud/core/pull/33994#issuecomment-479003389
